zerver/lib/message.py

```diff
@@ -1430,3 +1431,38 @@ def set_visibility_policy_possible(user_profile: UserProfile, message: Message)
 def remove_single_newlines(content: str) -> str:
     content = content.strip("\n")
     return re.sub(r"(?<!\n)\n(?!\n|[-*] |[0-9]+\. )", " ", content)
+
+
+def event_recipient_ids_for_action_on_messages(
+    messages: list[Message], *, exclude_long_term_idle_users: bool = False
+) -> set[int]:
+    assert len(messages) > 0
+    message_ids = [message.id for message in messages]
+
+    def get_user_ids_having_usermessage_row_for_messages(message_ids: list[int]) -> set[int]:
+        return set(
+            UserMessage.objects.filter(message_id__in=message_ids).values_list(
+                "user_profile_id", flat=True
+            )
+        )
+
+    sample_message = messages[0]
+    if not sample_message.is_stream_message():
+        return get_user_ids_having_usermessage_row_for_messages(message_ids)
+
+    channel_id = sample_message.recipient.type_id
+    channel = Stream.objects.get(id=channel_id)
+    subscriptions = get_active_subscriptions_for_stream_id(
+        channel_id, include_deactivated_users=False
+    )
+
+    if exclude_long_term_idle_users:
+        subscriptions = subscriptions.exclude(user_profile__long_term_idle=True)
+
+    subscriber_ids = set(subscriptions.values_list("user_profile_id", flat=True))
+
+    if channel.is_history_public_to_subscribers():
+        return subscriber_ids
+
+    user_ids_with_usermessage_row = get_user_ids_having_usermessage_row_for_messages(message_ids)
+    return user_ids_with_usermessage_row & subscriber_ids
```

Comment:
This logic differs from `has_message_access` in a key way

I would also place this next to `has_message_access` and make the docstrings/comments communicate that they are intended to implement the same logic.

Reply:
* Included the missing check
* Placed it next to `has_message_access`
* Added docstring + comments